### PR TITLE
New processor: urldecode

### PR DESCRIPTION
**Pull request by sincejune**
## What does this PR do?

This PR introduces a new processor `urldecode`. To decode fields from URL encoded format use this processor with `urldecode`.

## Why is it important?

This PR makes Filebeat can decode URL-encoded strings itself and then write to different outputs. 

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

**Comment by ycombinator** [comment]
Note to self (for review): Look at behavior of ES ingest node urldecode processor: https://www.elastic.co/guide/en/elasticsearch/reference/current/urldecode-processor.html.

File: libbeat/processors/actions/urldecode.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+func NewURLDecode(c *common.Config) (processors.Processor, error) {
+	config := urlDecodeConfig{
+		IgnoreMissing: false,
+		FailOnError:   true,
+	}
+
+	err := c.Unpack(&config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack the configuration of urldecode processor: %s", err)
+	}
```
Comment (ycombinator) [inline libbeat/processors/actions/urldecode.go:62]: Nit: you could shorten this to:

```suggestion
	if err := c.Unpack(&config); err != nil {
		return nil, fmt.Errorf("failed to unpack the configuration of urldecode processor: %s", err)
	}
```

File: libbeat/processors/actions/docs/urldecode.asciidoc [hunk 2]
```diff
@@ -0,0 +1,38 @@
+[[urldecode]]
+=== URL Decode
+
+++++
+<titleabbrev>urldecode</titleabbrev>
+++++
+
+The `urldecode` processor specifies a list of fields to decode from URL encoded format. Under the `fields`
+key, each entry contains a `from: old-key` and a `to: new-key` pair, where:
```
Comment (ycombinator) [inline libbeat/processors/actions/docs/urldecode.asciidoc:9]: How about calling them `source-field` and `target-field` instead of `old-key` and `new-key`, respectively? I think it makes the intent a bit clearer.

File: libbeat/processors/actions/docs/urldecode.asciidoc [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+* `from` is the original field name
+* `to` is the target field name
```
Comment (ycombinator) [inline libbeat/processors/actions/docs/urldecode.asciidoc:12]: Maybe add a note on this line about how this is optional and defaults to the `from` value?

File: libbeat/processors/actions/urldecode.go [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+
+	decodedData, err := url.QueryUnescape(encodedString)
+	if err != nil {
+		return fmt.Errorf("error trying to URL-decode %s: %v", encodedString, err)
+	}
+
+	target := to
+	// If to is empty
+	if to == "" || from == to {
```
Comment (ycombinator) [inline libbeat/processors/actions/urldecode.go:114]: Do we need the `from == to` clause?

**Comment by ycombinator** [comment]
I just noticed that this processor is placed under the `actions` package. Would you mind moving it to it's own package like we have done for some of the newer processors (see `add_id` or `fingerprint` as examples)?

File: libbeat/processors/actions/urldecode.go [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+	jsprocessor "github.com/elastic/beats/v7/libbeat/processors/script/javascript/module/processor"
+)
+
+type urlDecode struct {
+	config urlDecodeConfig
+	log    *logp.Logger
+}
+
+type urlDecodeConfig struct {
+	Fields        []fromTo `config:"fields"`
```
Comment (ycombinator) [inline libbeat/processors/actions/urldecode.go:40]:
```suggestion
	Fields        []fromTo `config:"fields" validate:"required"`
```

File: libbeat/processors/actions/urldecode.go (hunk 1 above)
Comment (sincejune) [reply libbeat/processors/actions/urldecode.go]: Updated in https://github.com/elastic/beats/pull/17505/commits/627ae785e9283b6a14e816c9fb4c627451b43ec5

File: libbeat/processors/actions/docs/urldecode.asciidoc (hunk 2 above)
Comment (sincejune) [reply libbeat/processors/actions/docs/urldecode.asciidoc]: Updated in https://github.com/elastic/beats/pull/17505/commits/627ae785e9283b6a14e816c9fb4c627451b43ec5

File: libbeat/processors/actions/docs/urldecode.asciidoc (hunk 3 above)
Comment (sincejune) [reply libbeat/processors/actions/docs/urldecode.asciidoc]: Updated in https://github.com/elastic/beats/pull/17505/commits/627ae785e9283b6a14e816c9fb4c627451b43ec5

File: libbeat/processors/actions/urldecode.go (hunk 4 above)
Comment (sincejune) [reply libbeat/processors/actions/urldecode.go]: Removed in https://github.com/elastic/beats/pull/17505/commits/627ae785e9283b6a14e816c9fb4c627451b43ec5

File: libbeat/processors/actions/urldecode.go (hunk 5 above)
Comment (sincejune) [reply libbeat/processors/actions/urldecode.go]: Added in https://github.com/elastic/beats/pull/17505/commits/627ae785e9283b6a14e816c9fb4c627451b43ec5

**Comment by sincejune** [comment]
Thanks @ycombinator , I moved `urldecode` processor into `libbeat/processors/urldecode/` in https://github.com/elastic/beats/pull/17505/commits/9e9457042d863b8a2fc0f3a0c1182ae8d998c1d9, but looks like the travis-ci didn't test the commit.

File: libbeat/processors/urldecode/urldecode.go [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+type urlDecodeConfig struct {
+	Fields        []fromTo `config:"fields" validate:"required"`
+	IgnoreMissing bool     `config:"ignore_missing"`
+	FailOnError   bool     `config:"fail_on_error"`
+}
+
+type fromTo struct {
+	From string `config:"from"`
```
Comment (ycombinator) [inline libbeat/processors/urldecode/urldecode.go:46]: Maybe add `validate:"required"` here?

File: libbeat/processors/urldecode/urldecode.go [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+func init() {
+	processors.RegisterPlugin("urldecode",
+		checks.ConfigChecked(NewURLDecode,
+			checks.RequireFields("fields"),
+			checks.AllowedFields("fields", "ignore_missing", "fail_on_error")))
+	jsprocessor.RegisterPlugin("URLDecode", NewURLDecode)
+}
+
+func NewURLDecode(c *common.Config) (processors.Processor, error) {
```
Comment (ycombinator) [inline libbeat/processors/urldecode/urldecode.go:58]: Nit: now that this is in the `urldecode` package, you can just name this `New`.

Comment (sincejune) [reply libbeat/processors/urldecode/urldecode.go]: Renamed in https://github.com/elastic/beats/pull/17505/commits/0347640dadb6e5fc7da04abd81b54b529b3fa738

File: libbeat/processors/urldecode/urldecode.go (hunk 6 above)
Comment (sincejune) [reply libbeat/processors/urldecode/urldecode.go]: Added in https://github.com/elastic/beats/pull/17505/commits/0347640dadb6e5fc7da04abd81b54b529b3fa738

**Comment by ycombinator** [comment]
CI failures are unrelated. Merging.

**Comment by sincejune** [comment]
Thanks for helping this, @ycombinator !
